File: backend/src/main/java/com/collabcrm/crdt/ORSet.java
package com.collabcrm.crdt;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * OR-Set (Observed-Remove Set) — CRDT für Mengen mit Add-Wins-Semantik.

 */
public class ORSet<T> {

    /**
     * Ein Dot identifiziert ein einzelnes Add-Event eindeutig.
     * Kombination aus nodeId (wer) und timestamp (wann) ist immer einzigartig.
     */
    public record Dot(String nodeId, long timestamp) {}


    public static class Entry<T> {
        public final T value;
        public final Set<Dot> addDots = new HashSet<>();     // Alle Add-Events
        public final Set<Dot> removeDots = new HashSet<>();   // Alle Remove-Events (beobachtete Dots)

        Entry(T value) {
            this.value = value;
        }


        boolean isAlive() {
            var surviving = new HashSet<>(addDots);
            surviving.removeAll(removeDots);
            return !surviving.isEmpty();
        }


        Entry<T> copy() {
            var e = new Entry<>(value);
            e.addDots.addAll(addDots);
            e.removeDots.addAll(removeDots);
            return e;
        }
    }


    private final Map<UUID, Entry<T>> entries = new ConcurrentHashMap<>();

    /**
     * Fügt ein neues Element mit zufälliger ID hinzu.
     * Erzeugt einen neuen Dot (nodeId + timestamp) als Add-Event.
     * Wird vom lokalen User aufgerufen wenn er ein Todo erstellt.
     */
    public UUID add(T element, String nodeId, long timestamp) {
        var id = UUID.randomUUID();
        var entry = new Entry<>(element);
        entry.addDots.add(new Dot(nodeId, timestamp));
        entries.put(id, entry);
        return id;
    }

    /**
     * Fügt ein Element mit expliziter ID hinzu (für Replikation).
     * Wird aufgerufen wenn eine Operation von einem anderen Node empfangen wird.
     * compute() ist atomar → kein Race Condition bei gleichzeitigen Operationen.
     */
    public void add(UUID elementId, T element, String nodeId, long timestamp) {
        entries.compute(elementId, (id, existing) -> {
            if (existing == null) {
                existing = new Entry<>(element);
            }
            existing.addDots.add(new Dot(nodeId, timestamp));
            return existing;
        });
    }


    public void remove(UUID elementId) {
        var entry = entries.get(elementId);
        if (entry != null) {
            entry.removeDots.addAll(entry.addDots);
        }
    }

    /**
     * Entfernt spezifische Dots (für Merge/Replikation).
     * Der Remote-Node teilt mit, welche Dots ER gesehen und entfernt hat.
     */
    public void remove(UUID elementId, Set<Dot> observedDots) {
        var entry = entries.get(elementId);
        if (entry != null) {
            entry.removeDots.addAll(observedDots);
        }
    }


    public Set<T> value() {
        var result = new HashSet<T>();
        for (var entry : entries.values()) {
            if (entry.isAlive()) {
                result.add(entry.value);
            }
        }
        return result;
    }


     // REST-Endpunkt  Todo-Liste eines Kunden zu liefern.

    public Map<UUID, T> elements() {
        var result = new LinkedHashMap<UUID, T>();
        for (var e : entries.entrySet()) {
            if (e.getValue().isAlive()) {
                result.put(e.getKey(), e.getValue().value);
            }
        }
        return result;
    }


    public void merge(ORSet<T> other) {
        for (var e : other.entries.entrySet()) {
            entries.compute(e.getKey(), (id, existing) -> {
                if (existing == null) {
                    return e.getValue().copy();
                }
                existing.addDots.addAll(e.getValue().addDots);
                existing.removeDots.addAll(e.getValue().removeDots);
                return existing;
            });
        }
    }

    // Für tests
    public ORSet<T> copy() {
        var result = new ORSet<T>();
        for (var e : entries.entrySet()) {
            result.entries.put(e.getKey(), e.getValue().copy());
        }
        return result;
    }


    public int size() {
        return (int) entries.values().stream().filter(Entry::isAlive).count();
    }

    public Map<UUID, Entry<T>> rawEntries() {
        return entries;
    }
}

File: frontend/src/crdt/ORSet.ts
/**
 * Dot — identifiziert ein einzelnes Add-Event eindeutig.
 */
export interface Dot {
  nodeId: string;
  timestamp: number;
}

/**
 * Entry — trackt Add- und Remove-Dots für ein Element.
 * Dots werden als serialisierte Strings "nodeId:timestamp" gespeichert
 * damit Set-Vergleiche korrekt funktionieren (JavaScript vergleicht Objekte by-reference).
 */
interface Entry<T> {
  value: T;
  addDots: Set<string>;    // serialisierte "nodeId:timestamp" Strings
  removeDots: Set<string>;
}

/** Serialisiert einen Dot zu einem eindeutigen String-Key */
function dotKey(d: Dot): string {
  return `${d.nodeId}:${d.timestamp}`;
}

/**
 * OR-Set (Observed-Remove Set) — TypeScript-Version.
 *
 * Identische Logik wie ORSet.java im Backend.
 * Verwendet für: Todo-Listen pro Kunde.
 *
 * Add-Wins-Semantik: Wenn User A ein Element entfernt und User B es gleichzeitig
 * hinzufügt, überlebt das Element — weil B's Dot nicht in A's Remove-Dots ist.
 */
export class ORSet<T> {
  // elementId → Entry mit allen Dots
  private entries = new Map<string, Entry<T>>();

  /** Fügt ein Element hinzu und erzeugt einen neuen Dot */
  add(elementId: string, value: T, nodeId: string, timestamp: number): void {
    let entry = this.entries.get(elementId);
    if (!entry) {
      entry = { value, addDots: new Set(), removeDots: new Set() };
      this.entries.set(elementId, entry);
    }
    entry.addDots.add(dotKey({ nodeId, timestamp }));
  }

  /**
   * Entfernt ein Element: Alle aktuellen Add-Dots werden zu Remove-Dots.
   * Neue Dots die gleichzeitig auf anderen Nodes entstehen überleben → Add-Wins.
   */
  remove(elementId: string): void {
    const entry = this.entries.get(elementId);
    if (entry) {
      for (const dot of entry.addDots) {
        entry.removeDots.add(dot);
      }
    }
  }

  /**
   * Prüft ob ein Element noch lebt:
   * Es lebt wenn mindestens ein Add-Dot existiert der nicht in Remove-Dots ist.
   */
  private isAlive(entry: Entry<T>): boolean {
    for (const dot of entry.addDots) {
      if (!entry.removeDots.has(dot)) return true;
    }
    return false;
  }

  /** Alle lebenden Elemente als Map (elementId → value) */
  get value(): Map<string, T> {
    const result = new Map<string, T>();
    for (const [id, entry] of this.entries) {
      if (this.isAlive(entry)) {
        result.set(id, entry.value);
      }
    }
    return result;
  }

  /** Anzahl der lebenden Elemente */
  get size(): number {
    let count = 0;
    for (const entry of this.entries.values()) {
      if (this.isAlive(entry)) count++;
    }
    return count;
  }

  /** Merged ein anderes OR-Set: Vereinigung aller Add- und Remove-Dots pro Element */
  merge(other: ORSet<T>): void {
    for (const [id, otherEntry] of other.entries) {
      let entry = this.entries.get(id);
      if (!entry) {
        entry = {
          value: otherEntry.value,
          addDots: new Set(),
          removeDots: new Set(),
        };
        this.entries.set(id, entry);
      }
      for (const dot of otherEntry.addDots) entry.addDots.add(dot);
      for (const dot of otherEntry.removeDots) entry.removeDots.add(dot);
    }
  }
}
